File: ruby/anagram/anagram.rb
class Anagram

  class ComparableWord

    attr_reader :comparable_sequence

    def initialize(word)
      @comparable_sequence = word.downcase.chars.sort
    end

    def anagram_of?(other)
      @comparable_sequence == other.comparable_sequence
    end

  end

  def initialize(word)
    @word = ComparableWord.new(word)
  end

  def match(candidates)
    candidates.select do |candidate|
      @word.anagram_of?(ComparableWord.new(candidate))
    end
  end

end

File: clojure/anagram/anagram.clj
(ns anagram)
(require '[clojure.string :as string])

(defn- comparable [word]
  (frequencies (string/lower-case word)))

(defn- anagram-matcher [word]
  (let [comparable-word (comparable word)]
    (fn [other-word]
      (= comparable-word (comparable other-word)))))

(defn anagrams-for [word candidates]
  (filter (anagram-matcher word) candidates))
